Fixes #18715 - support for fdi.script option

lzap opened the pull request:
New `fdi.script` option is added, it requies base64-encoded gzipped script which is executed during boot. In PXE-less mode (when BOOTIF is not present on kernel commandline) it is executed before NetworkManager is started, in PXE mode it is executed after NetworkManager finished configuration and IP is assigned.

To generate compatible string use:

    echo "echo 'This is a test'" | gzip -9 | base64 -w0
    H4sIAK4hZ1oCA0tNzshXUA/JyCxWAKJEhZLU4hJ1LgA92U9qFgAAAA==

and then append the option to kernel command line. See system journal for output (grep for `discovery-script`).

    APPEND ... fdi.script=H4sIAK4hZ1oCA0tNzshXUA/JyCxWAKJEhZLU4hJ1LgA92U9qFgAAAA==

lzap commented on 20-packages.ks:
```diff
@@ -66,9 +66,13 @@ unzip
 shim
+# These were renamed and split in EL 7.4+
+grub2-efi-x64
+grub2-efi-x64-cdboot
```
This line is actually the one which fixes missing EFI loader. All the rest is just documentation around building or the script feature.

lzap commented:
Here is a scratchbuild, won't last long, hurry up: http://downloads.theforeman.org/discovery/scratch/lzap/

stbenjam commented on root/usr/bin/discovery-script:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+basescript=${KCL_FDI_SCRIPT:-H4sIAOj1ZloCA0tNzshXUPLLVyhOLsosKFEoSCwuTk1R4gIAIbwTwhgAAAA=}
+echo "$basescript" | base64 -d | gunzip > /tmp/discovery-script
```
I think there should be a little more error checking here.  COMMAND_LINE_SIZE isn't very big: 2K on my Fedora x86 box.  A script end up truncated, and I know it's an extreme and contrived example but "rm -rf /tmp/foo" could end up as "rm -rf /":

```
echo "rm -rf /tmp/foo" | gzip -9 | base64 -w0 | cut -c 1-26 | base64 -d | gunzip
```

Which I know isn't a big deal on the discovery image, but base64 will return an error code if the input isn't quite valid, can we check it?

lzap replied:
Ok it looks like if a single character is missing (except `=` which is safe to cut) gunzip return non-zero code. I will take advantage and fail.

stbenjam commented:
This is pretty clever!

lzap commented:
Fixed, I added also a helper script which will encode file and show total length and the limit:

```
[lzap@box foreman-discovery-image]$ cat 20-packages.ks | aux/discovery-script-creator
fdi.script=H4sIANzQaVoCA3VV224cNwx911cIMAIkTeVtjKQNWhR9qdMYSJogbp4NjcSZIVY3iNRe8vWlZsZ2ErT74LUkkuI5PEf7pFi3txOQNgZOLjQPPjtSg6VZ7aEmCGrATB4OyUZQU23D1frXcM6BFFyNVGqeSBVLdPSq5IDu7HKFxigBbt67nEacVM2ZRwxA6tyiOmA0ERNGG5R1Bb3ikYsKwRfrFVKO/hVJnLrQf1brGuuIRJgm7aGQfjozF/p1txva9AVDsJcV/Gz50uW4ozkf7+Tg0k34B/rfX1y9fvX6xU/PFNuqJLz0qm+sY6hqXL+A5w5IJeANmY/owWUP0mplc5TiS5ogizZpAX06K2o+qw8F0s3H9zcP/2wVsiyJghrXFLOmCHnnCaKhaCvfLXt3HsnlA9TznfAxQb/nJklb0hseQD8cy0xO4Lby94USHPlhMVriuwmY4cRqP/iFPxAupk4dtVJyZSVDINVnodgV3+JCyIcUzjoBiAb0AM42Ah3keufXC3VtifSuUd0NmHYjVjjaEEwex4AJjIteqrQk4/bImJMcnn/TxbKb5eYhonTldSvEFWy81J8gCiqdkwNdIYAlORaYJLlS6Yg8a2SNpO3BYrCD9Ht/6wLs9vatts51MCvXs3EBIfHjmqBKxSVYhs9SKgSNnVsZvOjeNs7RMrrerJZJ6Ubqb+Bjrvv3NskwqmoNl+s+9/44CwJvGXTXhrbcaWEUc7S0Kes69Va1wMRShHaxUFrNsOpHf75+c7O7BdcqDGKKh7HAiH0dp7p5TDYUzRgl758ZZB5HqJ2rbkavbfKaSugUJX39Tv9y+fL5Y6I5/fzy25VxvpdfKi6nUnadLIu4Naxt8wx6UaFUJRZaoOoSmuhHLF5lgip666M63Rv/8QkYhIArRWfirm/OVaqYuLGoLv7zIz18vH+EhNtNExf/Fy3hnwT6W+n3WqZYKgor74TfU5cYOaG8S2/BpczXW181ov+CBBXdwuARu/aIViqUuBREzntJBhb1oLxPRkAbUcq8vWTmPmdz4tLU0ncnT1ynBxQNmu4/IzaqsT8jEtUtp1d+RB5tHJXxsVqRlwmHeNVDqN8uaHplN1uZbB71d4/qj+LLgWAFpJ8uizVLvLI5mOf8TMB89xobN4Pbr7sd4ZK1Rf2gFpmJYKnb7phb6A+BTij+lNGEnAmUGeUnoloT8pTpYbXZ16TM8r4/geTVv5JFDdNcBgAA
Maximum length: 512 (minus what's on kernel command line)
Actual length: 1243
Consider shell minifier or stage2 script downloaded from network.

[lzap@box foreman-discovery-image]$ echo "Test" | aux/discovery-script-creator
fdi.script=H4sIACHRaVoCAwtJLS7hAgDCGnj6BQAAAA==
Maximum length: 512 (minus what's on kernel command line)
Actual length: 47
```

stbenjam commented:
This is pretty sweet, I just built an image off the latest changes and it works perfectly.  The helper functions are nice, especially since the kernel cmdline is so limited. 

ACK from me

lzap commented:
Thanks, note for googlers - the example in the repo (configuring bonding) does send facts correctly, but Discovery currently errors out with "No IP address detected". This is because in the current design we need all interfaces to have IP address and MAC address, which is not the case for bonds. This is a RFE: http://projects.theforeman.org/issues/18583
